core/settlement/engine.go

```diff
@@ -289,9 +287,13 @@ func (e *Engine) SettleMTM(ctx context.Context, markPrice *num.Uint, positions [
 
 	for _, evt := range positions {
 		party := evt.Party()
-		// get the current position, and all (if any) position changes because of trades
-		current := e.getCurrentPosition(party, evt)
-		// we don't care if this is a nil value
+		// NB:
+		// current = last mtm'ed position if there is such of the position from the position event
+		// lastSettledPrice = last mark price if there is mtm'ed position or the price from the position event
+		// In addition if there is no previous mtm'ed position, e.settledPosition is updated with the position event size.
+		// @@EVODelavega this seems wrong to me - I think in case there was no previous mtm'ed position size should be initialised to 0
```

Comment:
IIRC, the `getCurrentPosition` would either get the party position, or create a new one with size 0 if needed (ie initialise the position to 0). At some point, in market.go `handleConfirmation` we registered the party/position with the settlement engine even if no trades had occurred, essentially creating a 0-initialised entry. 

As far as initialising the position with the event: because we continuously marked to market, any time we would hit this point, a new party that had just traded would've traded at the new mark price, so their position didn't need to be marked to market. That's why we could just use the positions engine event. Now, though, we should initialise to 0, and MTM all trades based on the price they were made at, compared to the new mark price

Reply:
well, no need to rely on memory here, just look at the old code - it's creating a new position with the size from the position engine's event. 

```
newPos(evt events.MarketPosition) *pos {
	return &pos{
		MarketPosition: evt,
		party:          evt.Party(),
		size:           evt.Size(),
		price:          evt.Price().Clone(),
	}
}
```

so it would not create a new one with 0, which is exactly my point. It creates a new one with the position engine's size which doesn't make sense to me.

Reply:
OK, a new party will have traded, that's a given. We skip the MTM settlement if they hold their position at the new mark price and if the  party hasn't traded. The current position size at current mark price here would be `mtmShare := delta.Mul(delta, num.NewUint(uint64(size)))`, or `mtmShare = 0 * size`. After that we mark the individual trades to market, same as before. I agree that a size of 0 to initialise makes more logical sense, but after looking at the current implementation a bit, I don't think it makes that much of a difference when marking to market